### ISSUE TITLE
Font loading proposal

### ISSUE THREAD

**Issue by sowasred2012** [canonical-web-and-design/vanilla-framework#2701]
## Background

- Large number of fonts being loaded when including Vanilla: [issue one][multiple-fonts-issue-one], [issue two][multiple-fonts-issue-two]

### Use `local()`

If a user happens to have the Ubuntu font already installed, we should leverage that. We can specify that the browser checks whether a given font exists locally before requesting it from a remote location, using `local()`. Safari on macOS requires the use of two `local()` rules to work, the first being the full name of the font, and the second being the Postscript name. In practice it would look like this:

```css
@font-face {
  font-display: $font-display-option;
  font-family: 'Ubuntu';
  font-style: normal;
  font-weight: 300;
  src:
    local('Ubuntu'),
    local('Ubuntu-Light'),
    url('#{$assets-path}46ed6870-Ubuntu-L-subset.woff2') format('woff2'),
    url('#{$assets-path}4070835e-Ubuntu-L-subset.woff') format('woff');
  unicode-range: U+0-FF, U+131, U+152, U+153, U+2BB, U+2BC, U+2C6, U+2DA, U+2DC, U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF, U+FFFD;
}
```

### Variable fonts

Variable fonts are a means of containing many, if not all, variants of a given font within a single file, offering the following benefits:

- Reduced number of network requests

  The big one, in terms of performance. By combining multiple font family variants into one file, we reduce the amount of network requests required to render fonts on a particular page from, [in our case][multiple-fonts-issue-one], at least [five][multiple-fonts-issue-two], to one.

  Also, although the size of a variable font file will necessarily be larger than a single standard font file, it will still be [smaller than the combined size][mandy-michael-slides] of the five or more standard font files currently in use.

- Increased font flexibility

  If it was decided that we needed a new variation of our font, rather than including a new font file for that purpose, we'd be able to define that variant directly in CSS.
  
- Fancy font animations

  Probably of little use to us, but it would open the door to doing things like [this][hello-demo].

Further reading:

- [MDN variable fonts article][mdn-article]
- [Microsoft variable fonts article & demo][ms-demo]
  
#### Creating a variable font file

Unfortunately it's not as simple as throwing all of our fonts into a web app; there are a number of font authoring tools out there, [quite][font-tool-1] [a few][font-tool-2] [of them][font-tool-3] seem to have the ability to generate variable fonts, so we would need to reach out to a font designer and have them create the file for us.

#### Browser compatibility

Variable fonts are relatively new, but browser support is [surprisingly good][vf-browser-support]. As mentioned in [this article][vf-article], we could use `@supports` to wrap variable fonts, and provide fallback fonts for older browsers (which will ignore `@supports` blocks), i.e. :

```css
@font-face {
  font-family: 'SourceSansVariable';
  src: url('source-sans-variable.woff2') format('woff2');
  font-weight: 1 999;
}

@font-face {
  font-family: 'SourceSans';
  src: url('source-sans-black.woff2') format('woff2'),
         url('source-sans-black.woff') format('woff');
  font-weight: 900;
}

@font-face {
  font-family: 'SourceSans';
  src: url('source-sans-semibold.woff2') format('woff2'),
         url('source-sans-semibold.woff') format('woff');
  font-weight: 600;
}

html {
  font-family: 'SourceSans' sans-serif;
}

@supports (font-variation-settings: normal) {
  html {
    font-family: 'SourceSansVariable', sans-serif;
  }
}
```

[multiple-fonts-issue-one]: https://github.com/canonical-web-and-design/vanilla-framework/issues/2368
[multiple-fonts-issue-two]: https://github.com/canonical-web-and-design/jaas.ai/issues/453
[vf-browser-support]: https://caniuse.com/#feat=variable-fonts
[vf-article]: https://medium.com/clear-left-thinking/how-to-use-variable-fonts-in-the-real-world-e6d73065a604
[hello-demo]: https://codepen.io/mandymichael/pen/dJZQaz
[font-tool-1]: https://help.fontlab.com/fontlab-vi/Variable-Fonts/
[font-tool-2]: https://glyphsapp.com/tutorials/creating-a-variable-font
[font-tool-3]: https://robofont.com/documentation/how-tos/creating-variable-fonts/
[mandy-michael-slides]: https://noti.st/mandy/AzhOpz#s6YhzCN
[mdn-article]: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Fonts/Variable_Fonts_Guide
[ms-demo]: https://developer.microsoft.com/en-us/microsoft-edge/testdrive/demos/variable-fonts/

**Comment by spencerbygraves** [comment]
WG: we will make this a Vanilla roadmap item for further investigation.

**Comment by bartaz** [comment]
We are addressing font issues separately.

Variable fonts would need to be discussed separately, as they affect Ubuntu font directly.

**Comment by barrymcgee** [comment]
It could be good to see this issue addressed again as @sowasred2012's investigation above seems valuable. 

A recent audit by @anthonydillon of www.ubuntu.com has identified fonts as a significant perf issue: https://www.fastorslow.com/app/profile/efd6c784-f500-5a70-9ec5-a9752436e62f

**Comment by bartaz** [comment]
Closed by #3135
